### PR TITLE
[LoRA][MoE] Fix PEFT 0.18+ target_parameters LoRA loading for 3D MoE experts (Qwen3.5)

### PR DESCRIPTION
**Pull request by vaderyang**
# Fix PEFT 0.18+ `target_parameters` LoRA loading for 3D MoE experts

## Summary

vLLM's runtime LoRA path (`--enable-lora --lora-modules`) silently drops every PEFT-trained adapter that uses `target_parameters` against fused 3D MoE expert tensors (e.g., Qwen3-MoE family). The model behaves like base — capbench scores `0.018` instead of the expected `0.68` for a known-good SFT, with trajectories degenerating into infinite tool-call loops.

This PR fixes two independent bugs that combine to break this path. Both are pure-Python; no kernel changes.

## Reproduction

- Base: `Qwen3-MoE` family (verified on Qwen3.5-35B-A3B / `Qwen3_5MoeForConditionalGeneration`).
- Adapter: PEFT 0.18+ trained with
  ```
  target_modules = "<regex matching q/k/v/o/gate/up/down_proj>"
  target_parameters = ["mlp.experts.gate_up_proj", "mlp.experts.down_proj"]
  ```
- Serve: `vllm serve <base> --enable-lora --lora-modules name=<adapter_path> --max-lora-rank 16`
- Observed: every inference behaves like base; benchmark scores collapse.
- Expected: matches PEFT `merge_and_unload` then serve (delta is mathematically equivalent).

## Bug A — `hf_to_vllm_mapper` is missing the LoRA path prefix on Qwen3.5

`Qwen3VLForConditionalGeneration.hf_to_vllm_mapper` rewrites base-weight keys via
`model.language_model.` → `language_model.model.`. PEFT adapters trained against
the language-only `Qwen3_5MoeForCausalLM` view save weights under the simpler
path `base_model.model.model.layers.X.mlp.experts.lora_*`. After the standard
`base_model.model.` strip those become `model.layers.X.mlp.experts.*` — no
`language_model.` segment for the existing rule to match — and the wrapped
multimodal class roots the language model at `language_model.model.layers.X.*`.
The lookup misses every expert layer; `_stack_moe_lora_weights` early-returns and
the LoRA buffer stays at zero.

**Fix:** override `hf_to_vllm_mapper` on `Qwen3_5ForConditionalGeneration` to add
`model.layers.` → `language_model.model.layers.`. The existing
`model.language_model.` rule still wins for base weights (longer prefix matches
first), so the new rule only kicks in for LoRA keys.

## Bug B — PEFT and vLLM use opposite `lora_A`/`lora_B` dim semantics for 3D MoE LoRA

PEFT 0.18's `ParamWrapper` (peft/tuners/lora/layer.py) treats a 3D parameter
`[E, dim1, dim2]` *positionally* as `[E, in=dim1, out=dim2]` and creates

```python
lora_A = nn.Linear(in_features=dim1, out_features=r * E)
lora_B = nn.Linear(in_features=r * E, out_features=dim2)
delta = einsum("o r e, e r i -> e i o", weight_B, weight_A)
```

For `gate_up_proj` of shape `[E, 2*intermediate, hidden]` the actual usage is

```python
y[..., e, 2*intermediate] = einsum("...d, eod -> ...eo", x_hidden, gate_up_proj)
```

so PEFT's "in" is the *output* side of the layer's natural forward and PEFT's
"out" is the *input* side.

vLLM's `add_lora_w13` / `add_lora_w2` punica kernels assume the opposite
(native) layout:

```
y_rank[e, r] = sum_in A[e, r, in]   * x[..., in]    # A applied to layer input
y_out [e, o] = sum_r  B[e, o,  r]   * y_rank[e, r]  # B produces layer output
```

So PEFT's `A` and `B` trade roles relative to vLLM's, and each tensor needs to
be transposed in its trailing two dims after the swap.

The complication: `LoRALayerWeights.create_dummy_lora_weights` (used during
warmup/CUDA-graph capture) builds dummy LoRAs in vLLM-native layout. The dummy
flat shape is identical to PEFT's flat shape but with opposite dim semantics, so
a unconditional swap+transpose breaks warmup before the real adapter ever loads.

**Fix:** detect the layout in `_stack_moe_lora_weights` and branch on it. The
trailing dim of `lora_a` is the discriminator: `hidden_size` → vLLM-native
(dummy LoRA), anything else → PEFT-native (real adapter, dim is
`2*intermediate`). The vLLM-native branch keeps the original code path. The
PEFT branch swaps and transposes:
- `vllm_w13_a = peft_B_w13^T` (applied to hidden), `vllm_w13_b = peft_A_w13^T` (produces 2*intermediate)
- `vllm_w2_a  = peft_B_w2^T`  (applied to intermediate), `vllm_w2_b  = peft_A_w2^T` (produces hidden)

Mathematically:

```
PEFT delta[e, j, k]      = sum_r peft_B[k, r, e] * peft_A[e, r, j]
output_lora[..., e, j]   = sum_k x[..., k] * delta[e, j, k]
                        = sum_r peft_A[e, r, j] * (sum_k peft_B[e, k, r] * x[..., k])
```

After the swap+transpose, vLLM's two-step kernel computes exactly this sum.
Down-proj follows the same derivation.

## Verification

End-to-end on Qwen3.5-35B-A3B + a known SFT LoRA (`r=16, alpha=32`) over 27
capbench dev tasks × 3 repeats:

| Path | overall_avg | strict_pass |
|------|-------------|-------------|
| LoRA via vLLM 0.19 / 0.20.0 (broken) | **0.018** | 1/27 |
| `merge_and_unload` then serve (baseline) | **0.681** | 20/27 |
| **LoRA via vLLM 0.20.0 + this PR** | **0.700** | 19/27 |

The patched LoRA path matches the merged baseline within run-to-run noise
(temperature=0 but vLLM batching introduces minor nondeterminism).

## Scope

- Tested with: Qwen3.5-35B-A3B + PEFT 0.18.1 adapter, single-GPU TP=1.
- Bug A is Qwen3.5-specific. Other multimodal MoE models that wrap a CausalLM
  language model under `language_model.` will need the same prefix rule if
  they accept PEFT adapters trained against the language-only view.
- Bug B is generic for any model that sets `is_3d_moe_weight = True` and is
  fine-tuned with PEFT `target_parameters` against fused expert tensors.

## Files Changed

- `vllm/lora/model_manager.py` — layout-detection branch in `_stack_moe_lora_weights`
- `vllm/model_executor/models/qwen3_5.py` — extended `hf_to_vllm_mapper`

## Notes for Reviewers

1. The layout detection (`shape[-1] == hidden_size`) is robust on Qwen3-MoE
   because `hidden` and `2*intermediate` are not equal in the public configs
   (e.g., Qwen3.5-35B-A3B has `hidden=2048, 2*intermediate=1024`). If a future
   MoE config has `hidden == 2 * intermediate`, the discriminator would
   degenerate; an explicit `lora_request.is_peft_target_parameters` flag could
   be threaded through if that becomes a concern.
2. `PEFTHelper` (`vllm/lora/peft_helper.py`) does not currently parse the
   `target_parameters` field from `adapter_config.json` — silent drop. Loading
   doesn't depend on it (vLLM matches by tensor key naming), so the field is
   informational only. Adding a future warning for unsupported
   target_parameters values would be a clean follow-up.
3. The existing `experts.base_layer` warning in `worker_manager.py:154`
   ("LoRA module ... is not in the model's supported LoRA target modules ...
   These parameters will be ignored") is a misleading log line for this case —
   the parameters *are* applied via `_stack_moe_lora_weights`. Could be
   refined separately.

### CODE REVIEW

**Comment by vaderyang** [comment]
Thanks for the careful review! Pushed v2 (force-pushed branch with the same 2 logical commits, refactored).

Changes since v1:
- **Heuristic fragility (your #3)** addressed: `PEFTHelper` now actually parses `target_parameters` from `adapter_config.json` (was previously silently dropped). `LoRAModel` carries it through, and `_stack_moe_lora_weights` uses it as the **primary** signal. The shape-based check is now the **fallback** for paths that build a `LoRAModel` without going through `PEFTHelper` (e.g. dummies, future callers). On configs where `hidden == 2 * intermediate` the explicit `target_parameters` list wins. Helper: `_target_parameters_indicates_3d_lora`.
- **Inline import** moved from class body to top-level `from .utils import (..., WeightsMapper, ...)`.
- All `pre-commit run` hooks pass on Python 3.10 (ruff check / ruff format / typos / mypy / SPDX / lazy-import / config validation).

Re: your **#1 (TP > 1)** — I think this one doesn't apply for the configurations vLLM currently supports for `FusedMoEWithLoRA`:

```python
# vllm/lora/layers/fused_moe.py:34-36
assert not self.base_layer.use_ep, (
    "EP support for Fused MoE LoRA is not implemented yet."
)
```

With `use_ep=False`, every TP rank holds the full set of experts (only `intermediate_size_per_partition` and the hidden axis get TP-sharded), so `module.w13_lora_a_stacked[0].shape[1] == local_num_experts == global_num_experts`, matching the `[E, dim1, dim2]` first axis of the PEFT-saved adapter — no expert-axis slicing needed in `_stack_moe_lora_weights`.

The TP>1 partition slicing along the hidden / 2\*intermediate / intermediate axes is then handled inside `FusedMoE3DWithLoRA.set_lora` (`_slice_w13_a/b`, `_slice_w2_a/b`). After the v2 swap+transpose, the tensors handed to `set_lora` have exactly the dim layout those `_slice_*` helpers expect (`[E, r, hidden]` for `w13_a`, `[E, 2*intermediate, r]` for `w13_b`, etc.), so the existing per-rank slicing applies unchanged. End-to-end I tested only TP=1, but the shape contract for TP>1 is the same as the pre-existing `vllm-native` (dummy) path that this PR doesn't change.

If `use_ep=True` ever gets supported for fused MoE LoRA, expert-dim slicing will need to be added — but that's a follow-up that affects the existing path equally.

Re: your **#2 (key naming)** — I think this is a misread of the existing code: the lookup `module_name + ".base_layer"` for gate_up and the bare `module_name` for down_proj is what PEFT 0.18+ actually saves. I verified on a real `peft==0.18.1` adapter that the on-disk safetensor keys are exactly:

```
base_model.model.<module_name>.experts.base_layer.lora_A.weight
base_model.model.<module_name>.experts.base_layer.lora_B.weight
base_model.model.<module_name>.experts.lora_A.weight
base_model.model.<module_name>.experts.lora_B.weight
```

— `experts.base_layer.lora_*` is gate_up (PEFT picks this name because the 3D parameter `gate_up_proj` is a sibling of an existing `down_proj` parameter and `ParamWrapper` happens to wrap it under `base_layer`); `experts.lora_*` (no `.base_layer.`) is down_proj. There's no `.gate_up_proj` / `.down_proj` suffix on the LoRA keys themselves.

(Happy to take any further suggestions — the ergonomics of having two related-but-distinct conventions for these expert LoRAs is worth a pass at some point.)

**Comment by vaderyang** [comment]
@HollowMan6 @jeejeelee — pinging since this builds directly on the `experts.base_layer` 3D MoE LoRA loading you've worked on (#37114, #36976). Could you take a look when you get a chance? PR is gated by the contributor `pre-run-check` (no merged PRs from me yet), so it's currently waiting on a `ready`/`verified` label to even run CI. End-to-end repro + math derivation are in the description; v2 follow-up addressed the gemini-bot review.

**Review by jeejeelee — CHANGES_REQUESTED** [review]
IIRC, perf 0.18 has a bug, which can't generate correct moe lora weight.
You also can refer to https://github.com/vllm-project/vllm/pull/42242
